### PR TITLE
dts: stm32: uart clocks

### PR DESCRIPTION
**Pull request by erwango**
Provide rcc node to stm32 based socs.
Use rcc node to provide clock information on STM32 u(s)art nodes

### CODE REVIEW

File: dts/arm/st/clocks.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ */
+#ifndef __DT_BINDING_ST_CLOCKS_H
+#define __DT_BINDING_ST_CLOCKS_H
+
+
+#define STM32_CLOCK_BUS_AHB1    0
+#define STM32_CLOCK_BUS_AHB2    1
+#define STM32_CLOCK_BUS_APB1    2
+#define STM32_CLOCK_BUS_APB2    3
+#ifdef CONFIG_SOC_SERIES_STM32L4X
```
Comment (galak) [inline dts/arm/st/clocks.h]: remove the #ifdef here.

File: include/drivers/clock_control/stm32_clock_control.h [hunk 2]
```diff
@@ -19,10 +19,10 @@ enum {
 	STM32_CLOCK_BUS_AHB1,
 	STM32_CLOCK_BUS_AHB2,
 	STM32_CLOCK_BUS_APB1,
+	STM32_CLOCK_BUS_APB2,
 #ifdef CONFIG_SOC_SERIES_STM32L4X
 	STM32_CLOCK_BUS_APB1_2,
 #endif
-	STM32_CLOCK_BUS_APB2,
 };
```
Comment (galak) [inline include/drivers/clock_control/stm32_clock_control.h]: remove this enum, and just the defines from dts/arm/st/clocks.h.

File: dts/arm/st/stm32f103Xb.dtsi [hunk 3]
```diff
@@ -32,6 +32,14 @@
 	};
 
 	soc {
+		rcc: rcc@40021000 {
+			compatible = "st,stm32-rcc";
+			clocks-controller;
+			#clocks-cells = <2>;
+			reg = <0x40021000 0x400>;
+			label = "STM32_CLOCK";
```
Comment (galak) [inline dts/arm/st/stm32f103Xb.dtsi]: is the label here for code generation purposes?  If so how about STM32_CLK_RCC

**Review by galak — CHANGES_REQUESTED** [review]
Some cleanup on enum / defines, but otherwise looks fine.

File: dts/arm/yaml/st,stm32-rcc.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  - reg:
+      type: array
+      description: mmio register space
+      generation: define
+      category: required
+
+cell_string: clocks
+
+"#cells":
+  - bus
```
Comment (galak) [inline dts/arm/yaml/st,stm32-rcc.yaml]: is this correct, should we have:

- phandle
- bus
- bits

?

File: dts/common/yaml/uart.yaml [hunk 5]
```diff
@@ -27,4 +27,14 @@ properties:
       category: required
       description: Human readable string describing the device (used by Zephyr for API name)
       generation: define
+  - interrupts-name:
+      type: array
```
Comment (ydamigos) [inline dts/common/yaml/uart.yaml]: Do you mean "interrupt-names"?
The interrupt-names property consists of a list of strings so the "stringlist" should be used here.

File: dts/arm/st/stm32f103Xb.dtsi (hunk 3 above)
Comment (erwango) [inline dts/arm/st/stm32f103Xb.dtsi]: arbitrary choice, STM32_CLK_RCC is ok

File: dts/arm/yaml/st,stm32-rcc.yaml (hunk 4 above)
Comment (erwango) [inline dts/arm/yaml/st,stm32-rcc.yaml]: I can have a try but I don't know if script will handle it.
It's a bit picky with clocks node (I'll fw you a mail about that)

File: dts/common/yaml/uart.yaml (hunk 5 above)
Comment (erwango) [inline dts/common/yaml/uart.yaml]: Indeed, thanks!

File: include/drivers/clock_control/stm32_clock_control.h (hunk 2 above)
Comment (erwango) [inline include/drivers/clock_control/stm32_clock_control.h]: I'm able to #include <cloks.h>
I think it'd be better to #include <dts/arm/st/clocks.h>, but I can't get it work
Any idea how I could do that?

File: dts/arm/yaml/st,stm32-rcc.yaml (hunk 4 above)
Comment (erwango) [inline dts/arm/yaml/st,stm32-rcc.yaml]: seems to work, let's go

File: include/drivers/clock_control/stm32_clock_control.h (hunk 2 above)
Comment (erwango) [inline include/drivers/clock_control/stm32_clock_control.h]: So best I could do is #include <arm/st/clocks.h>
I'll change the plan and #include <dt_stm32_clocks.h>

**Comment by ydamigos** [comment]
According to linux device tree bindings for "st,stm32-rcc" (http://elixir.free-electrons.com/linux/v4.13-rc2/source/Documentation/devicetree/bindings/clock/st,stm32-rcc.txt):

> - #clock-cells: 2, device nodes should specify the clock in their "clocks"
>   property, containing a phandle to the clock device node, an index selecting
>   between gated clocks and other clocks and an index specifying the clock to
> use.

@erwango Do we plan to follow linux DT bindings for clocks or use something different for Zephyr?

**Comment by erwango** [comment]
@ydamigos : Thanks for raising this question.
Long term plan is indeed to converge with Linux DT (not "converge to" Linux DT), but we don't want this to block zephyr short term objectives/progress, and hence I'd like to  be able to derogate when needed.

There are several reasons to that:
1-Using Linux bindings may require to either rework zephyr driver either to have binding not reflecting
zephyr implementation of being over complicated vs zephyr driver need. 
For instance on rcc node, Linux biding is the output of a complex driver (clock gating, ability to have clock per IP, ..) which is not in line with the current rcc driver on Zephyr. Having same bindings on Zephyr would be nice but would not reflect current Zephyr RCC driver.

2-"Linux time" is different from "Zephyr time", they have their own objectives and plan. Linux STM32 device tree is not frozen and evolutions are on-going. Given Linux upstream process, it can take time and may require several reworks (hence months), before we know on what we should align on.

3-Zephyr will have more stm32 dts files than Linux. If we want to have a dts collection of files coherent from {zephyr, Linux, ..} global point of view, Linux dts will have to accept choices that might not be
coherent from single Linux point of view.

This being said, if there are options to get things closer to Linux, or more coherent from global point of view, while keeping things simple in zephyr and not preventing us to progress at our expected pace, we should use them.

File: dts/arm/Makefile [hunk 6]
```diff
@@ -1,4 +1,9 @@
 ifeq ($(CONFIG_HAS_DTS),y)
+
+ifdef CONFIG_SOC_FAMILY_STM32
+ZEPHYRINCLUDE += -I$(srctree)/dts/arm/st
```
Comment (galak) [inline dts/arm/Makefile]: Can we not just include via:

#include <dts/arm/st/dt_stm32_clocks.h>

File: dts/arm/st/dt_stm32_clocks.h [hunk 7]
```diff
@@ -0,0 +1,17 @@
+/*
```
Comment (galak) [inline dts/arm/st/dt_stm32_clocks.h]: any reason not to just name this stm32_clocks.h (related to how we include it)

File: dts/arm/st/dt_stm32_clocks.h [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+ */
+#ifndef __DT_BINDING_ST_CLOCKS_H
+#define __DT_BINDING_ST_CLOCKS_H
+
+
+#define STM32_CLOCK_BUS_AHB1    0
+#define STM32_CLOCK_BUS_AHB2    1
+#define STM32_CLOCK_BUS_APB1    2
+#define STM32_CLOCK_BUS_APB2    3
+#define	STM32_CLOCK_BUS_APB1_2  4
```
Comment (galak) [inline dts/arm/st/dt_stm32_clocks.h]: some odd whitespace issue, guessing a tab after define

File: include/drivers/clock_control/stm32_clock_control.h [hunk 9]
```diff
@@ -10,21 +10,11 @@
 #define _STM32_CLOCK_CONTROL_H_
 
 #include <clock_control.h>
+#include <dt_stm32_clocks.h>
```
Comment (galak) [inline include/drivers/clock_control/stm32_clock_control.h]: lets include this via #include <dts/arm/st/dt_stm32_clocks.h>

**Review by galak — CHANGES_REQUESTED** [review]
Some minor nits and how we access the includes

File: dts/arm/Makefile (hunk 6 above)
Comment (erwango) [inline dts/arm/Makefile]: I had a try but didn't work. Maybe I was not lucid anymore but is there any subtlety to achieve this?
I guess #include "dts/arm/st/dt_stm32_clocks.h" would be more likely to work

File: dts/arm/st/dt_stm32_clocks.h (hunk 7 above)
Comment (erwango) [inline dts/arm/st/dt_stm32_clocks.h]: Indeed I added dt_ to compensate the missing path

**Comment by erwango** [comment]
@galak: Are you good with new version?

**Review by galak — APPROVED** [review]

**Pull request merged**